contracts/sg-ics721/src/execute.rs

```diff
@@ -113,4 +115,70 @@ impl Ics721Execute for SgIcs721Contract {
 
         to_json_binary(&instantiate_msg)
     }
+
+    fn mint_msg(
+        &self,
+        token_id: String,
+        token_uri: Option<String>,
+        owner: String,
+        data: Option<Binary>,
+    ) -> StdResult<Binary> {
+        // parse token data and check whether it is of type NftExtension
+        let extension: Metadata = match data {
+            Some(data) => {
+                match from_json::<NftExtension>(data).ok().map(|ext| Metadata {
+                    animation_url: ext.animation_url,
+                    attributes: ext.attributes.map(|traits| {
+                        traits
+                            .into_iter()
+                            .map(|t| Trait {
+                                trait_type: t.trait_type,
+                                value: t.value,
+                                display_type: t.display_type,
+                            })
+                            .collect()
+                    }),
+                    background_color: ext.background_color,
+                    description: ext.description,
+                    external_url: ext.external_url,
+                    image: ext.image,
+                    image_data: ext.image_data,
+                    youtube_url: ext.youtube_url,
+                    name: ext.name,
+                }) {
+                    Some(extension) => extension,
+                    None => Metadata {
+                        animation_url: None,
+                        attributes: None,
+                        background_color: None,
+                        description: None,
+                        external_url: None,
+                        image: None,
+                        image_data: None,
+                        youtube_url: None,
+                        name: None,
+                    },
+                }
+            }
+            None => Metadata {
+                animation_url: None,
+                attributes: None,
+                background_color: None,
+                description: None,
+                external_url: None,
+                image: None,
+                image_data: None,
+                youtube_url: None,
+                name: None,
+            },
+        };
+
+        let msg = sg721_metadata_onchain::ExecuteMsg::Mint {
```

Comment:
so my question here is, does every meta data load resolve into a metadata_onchain type message? Wondering if there is a different struct for off-chain metadata?

Reply:
In case of offchain metadata:
- extension is empty onchain Metadata (with `None` values for each prop like image and attributes)
- `token_uri` is used for offchain data, e.g. refers to an IPFS link with offchain metadata. Here JSON file on IPFS is exactly the same type for `Metadata` (as defined by ERC721).

So SG NFT mint is done by this:

```rust
        let msg = cw721_metadata_onchain::msg::ExecuteMsg::Mint {
            token_id,
            token_uri, // holds off-chain metadata
            owner,
            extension, // holds on-chain metadata
        };
```

Reply:
Just wondering here, if off chain metadata is minted in this way, shouldn't we not use an execute message from cw721_onchain such as to not confuse the developer? I think the data is fine just coming from the cw721_metadata_onchain directory might be a bit misleading.

Reply:
Oh my bad (wrong example given above). This is an sg-implementation, proper code is using this:

```rust
        let msg = sg721_metadata_onchain::ExecuteMsg::Mint {
            token_id,
            token_uri, // holds off-chain metadata
            owner,
            extension, // holds on-chain metadata
        };
        to_json_binary(&msg)
```

Please note that `mint_msg()` returns `StdResult<Binary>` for ics721 and sg-ics721.